### PR TITLE
Feat/reset offset

### PR DESCRIPTION
**Pull request by Debatent**

### CODE REVIEW

**Comment by Debatent** [comment]
#61

File: api/src/test/java/com/michelin/ns4kafka/controllers/ConsumerGroupControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        when(consumerGroupService.getPartitionsToReset(ns, "groupID", "topic1"))
+                .thenReturn(topicPartitions);
+        when(consumerGroupService.prepareOffsetsToReset(ns, "groupID", null, topicPartitions, ResetOffsetsMethod.TO_EARLIEST))
+                .thenReturn(Map.of(topicPartition1, 5L, topicPartition2, 10L));
+
+
+        ConsumerGroupResetOffsets result = consumerGroupController.resetOffsets("test", "groupID", resetOffset, true);
+        verify(consumerGroupService, never()).alterConsumerGroupOffsets(notNull(), anyString(), anyMap());
+        assertTrue(result.getStatus().isSuccess());
+        assertEquals(result.getStatus().getOffsetChanged().get(topicPartition1.toString()), 5L);
```
Comment (twobeeb) [inline api/src/test/java/com/michelin/ns4kafka/controllers/ConsumerGroupControllerTest.java:125]: Expected goes first

File: api/src/test/java/com/michelin/ns4kafka/controllers/ConsumerGroupControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+                        .build())
+                .build();
+
+        when(namespaceService.findByName("test"))
+                .thenReturn(Optional.of(ns));
+        when(consumerGroupService.validateResetOffsets(resetOffset))
+                .thenReturn(new ArrayList<>());
+        when(consumerGroupService.isNamespaceOwnerOfConsumerGroup("test", "groupID"))
+                .thenReturn(false);
+        assertThrows(ResourceValidationException.class,
```
Comment (twobeeb) [inline api/src/test/java/com/michelin/ns4kafka/controllers/ConsumerGroupControllerTest.java:192]: Validate the ResourceValidationException message

File: api/src/test/java/com/michelin/ns4kafka/controllers/ConsumerGroupControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                .build();
+
+        when(namespaceService.findByName("test"))
+                .thenReturn(Optional.of(ns));
+        when(consumerGroupService.validateResetOffsets(resetOffset))
+                .thenReturn(List.of("Validation Error"));
+        when(consumerGroupService.isNamespaceOwnerOfConsumerGroup("test", "groupID"))
+                .thenReturn(true);
+        assertThrows(ResourceValidationException.class,
+                () -> consumerGroupController.resetOffsets("test", "groupID", resetOffset, false));
```
Comment (twobeeb) [inline api/src/test/java/com/michelin/ns4kafka/controllers/ConsumerGroupControllerTest.java:223]: Validate the ResourceValidationException message

File: api/src/test/java/com/michelin/ns4kafka/controllers/ConsumerGroupControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        when(namespaceService.findByName("test"))
+                .thenReturn(Optional.of(ns));
+        when(consumerGroupService.validateResetOffsets(resetOffset))
+                .thenReturn(new ArrayList<>());
+        when(consumerGroupService.isNamespaceOwnerOfConsumerGroup("test", "groupID"))
+                .thenReturn(true);
+        when(consumerGroupService.getConsumerGroupStatus(ns, "groupID"))
+                .thenReturn("Active");
+        assertThrows(ResourceValidationException.class,
+                () -> consumerGroupController.resetOffsets("test", "groupID", resetOffset, false));
```
Comment (twobeeb) [inline api/src/test/java/com/michelin/ns4kafka/controllers/ConsumerGroupControllerTest.java:254]: Validate the ResourceValidationException message

File: api/src/test/java/com/michelin/ns4kafka/services/ConsumerGroupServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    void doValidation_MissingTopic() {
+        ConsumerGroupResetOffsets consumerGroupResetOffsets = ConsumerGroupResetOffsets.builder()
+                .spec(ConsumerGroupResetOffsetsSpec.builder()
+                        .topic(":2")
+                        .method(ResetOffsetsMethod.TO_EARLIEST)
+                        .build())
+                .build();
+
+        List<String> result = consumerGroupService.validateResetOffsets(consumerGroupResetOffsets);
+        assertTrue(result.size() == 1);
```
Comment (twobeeb) [inline api/src/test/java/com/michelin/ns4kafka/services/ConsumerGroupServiceTest.java:82]: use assertEquals

File: api/src/test/java/com/michelin/ns4kafka/services/ConsumerGroupServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+        ConsumerGroupResetOffsets consumerGroupResetOffsets = ConsumerGroupResetOffsets.builder()
+                .spec(ConsumerGroupResetOffsetsSpec.builder()
+                        .topic("namespace_testTopic01:2")
+                        .method(ResetOffsetsMethod.TO_DATETIME)
+                        .options("2021-06-025T11:23:33.0249+02:00")
+                        .build())
+                .build();
+
+        List<String> result = consumerGroupService.validateResetOffsets(consumerGroupResetOffsets);
+        assertTrue(result.size() == 1);
```
Comment (twobeeb) [inline api/src/test/java/com/michelin/ns4kafka/services/ConsumerGroupServiceTest.java:138]: use assertEquals

File: api/src/test/java/com/michelin/ns4kafka/services/ConsumerGroupServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+        when(applicationContext.getBean(ConsumerGroupAsyncExecutor.class,
+                Qualifiers.byName(namespace.getMetadata().getCluster()))).thenReturn(consumerGroupAsyncExecutor);
+        try {
+            when(consumerGroupAsyncExecutor.getCommittedOffsets(anyString())).thenReturn(
+                    Map.of(new TopicPartition("topic1", 0), 10L,
+                            new TopicPartition("topic1", 1), 15L,
+                            new TopicPartition("topic2", 0), 10L));
+        } catch (Exception e) {
+            assertTrue(false);
+        }
```
Comment (twobeeb) [inline api/src/test/java/com/michelin/ns4kafka/services/ConsumerGroupServiceTest.java:249]: don't do this

File: api/src/test/java/com/michelin/ns4kafka/services/ConsumerGroupServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+
+        Map<TopicPartition, Long> result = null;
+        try {
+            result = consumerGroupService.prepareOffsetsToReset(namespace, groupId, options, partitionsToReset, method);
+        } catch (Exception e) {
+            assertTrue(false);
+        }
```
Comment (twobeeb) [inline api/src/test/java/com/michelin/ns4kafka/services/ConsumerGroupServiceTest.java:256]: don't do this, call normally

File: cli/src/main/java/com/michelin/ns4kafka/cli/services/ConsumerGroupService.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    @Inject
+    public ConsumerGroupClient consumerGroupClient;
+    @Inject
+    public LoginService loginService;
+
+
+    public Resource reset(String namespace, String group, Resource resource, boolean dryRun) {
+        return consumerGroupClient.reset(loginService.getAuthorization(), namespace, group, resource, dryRun);
+    }
+}
```
Comment (twobeeb) [inline cli/src/main/java/com/michelin/ns4kafka/cli/services/ConsumerGroupService.java:22]: Integrate this in ResourceService

File: cli/src/main/java/com/michelin/ns4kafka/cli/client/ConsumerGroupClient.java [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+    @Post("{namespace}/consumer-groups/{consumerGroupName}/reset{?dryrun}")
+    Resource reset(
+            @Header("Authorization") String token,
+            String namespace,
+            String consumerGroupName,
+            @Body Resource json,
+            @QueryValue boolean dryrun);
+
+}
```
Comment (twobeeb) [inline cli/src/main/java/com/michelin/ns4kafka/cli/client/ConsumerGroupClient.java:18]: Integrate this in NamespacedResourceClient

**Review by twobeeb — CHANGES_REQUESTED** [review]

**Pull request merged**